Added proposal for auto-rebalance on imbalanced cluster feature in operator

ShubhamRwt opened the pull request:
This PR aims to introduce the self-healing feature in Strimzi. This proposal contains all the comments and suggestion left on the old proposal #145. This proposal aim to utilize the `auto-rebalancing` feature of Strimzi to introduce the self healing.

ShubhamRwt committed:
Added proposal for self-healing

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+
+The above flow diagram depicts the self-healing process in Cruise Control.
+The anomaly detector manager detects an anomaly (using the detector classes) and forwards it to the notifier.
+The notifier classes like `SelfHealingNotifier`, `AlertaSelfHealingNotifier`, `SlackSelfHealingNotifier` etc. provides alerts to the users about the detected anomaly and also returns the action that needs to be taken on the anomaly i.e. whether to fix it, ignore it or delay it.
```
I would not mention the classes again but just "the configured notifier"

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+The frequency of this check can be changed via the `anomaly.detection.interval.ms` configuration.
+Detector classes have different mechanisms to detect their corresponding anomalies.
+For example, `KafkaBrokerFailureDetector` utilises Kafka Metadata API whereas `DiskFailureDetector` and `TopicAnomalyDetector` utilises Kafka Admin API.
+Furthermore, `MetricAnomalyDetector` and `GoalViolationDetector` use metrics to detect their anomalies.
```
Is goal violation using metrics?

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+For example, `KafkaBrokerFailureDetector` utilises Kafka Metadata API whereas `DiskFailureDetector` and `TopicAnomalyDetector` utilises Kafka Admin API.
+Furthermore, `MetricAnomalyDetector` and `GoalViolationDetector` use metrics to detect their anomalies.
+The detected anomalies can be of various types:
+* Goal Violation - This happens if certain optimization goals are violated (e.g. DiskUsageDistributionGoal etc.). These goals can be configured through the `self.healing.goals` configuration under the `spec.cruiseControl.config` section. These goals are independent of the manual rebalancing goals configured in the `KafkaRebalance` custom resource.
```
But we don't allow to set self.healing stuff, so?

scholzj replied:
So is it allowed or is it not allowed? I read this as that this is already allowed and supported.

scholzj replied:
So maybe what you want to say is something like `These goals can be configured through the `self.healing.goals` option in Cruise Control configuration. However, this option is forbidden in the `spec.cruiseControl.config` section.

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+The detected anomalies can be of various types:
+* Goal Violation - This happens if certain optimization goals are violated (e.g. DiskUsageDistributionGoal etc.). These goals can be configured through the `self.healing.goals` configuration under the `spec.cruiseControl.config` section. These goals are independent of the manual rebalancing goals configured in the `KafkaRebalance` custom resource.
+* Topic Anomaly - Where one or more topics in cluster violates user-defined properties (e.g. some partitions are too large in disk).
+* Broker Failure - This happens when a non-empty broker crashes or leaves a cluster.
```
"leaves a cluster" but for too long that's considered an error. The failure is not detected immediately everytime you do a scale down for example.

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+### Proposal
+
+This proposal allows the users to have their cluster balanced automatically whenever an anomaly is detected by Cruise Control.
+To ensure that the operator drives the rebalance we will be leveraging the Cruise Control self-healing mechanism but not completely.
```
We are not leveraging the self-healing at all, we are leveraging the anomaly detection only. We need to be careful about mentioning self-healing here. As said before people could get confused if we are using or not (which is the case) the CC self-healing feature.

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+
+#### `Full` mode in auto-rebalancing feature of Strimzi
+
+To leverage the above self-healing mechanism, we will be introducing a new mode to the `auto-rebalancing` feature.
```
Still self-healing :-)

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+Once the auto-rebalance with `Full` mode is enabled, self-healing will be enabled and a notifier `AnomalyDetectorNotifier` will be set(default notifier).
+This notifier's job will be to update the operator regarding the anomalies being detected so that the operator can trigger a rebalance.
+The user can change the default notifier if they want by setting the `anomaly.notifier.class` to point towards their own custom notifier.
```
But if they are going to do that, they will miss the auto fix from Strimzi.
Also it seems you are proposing to configure our own notifier if the full auto-rebalancing is set but we should clarify that a manual user configuration would override it.

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+#### Using other notifiers when using `Full` mode
+
+Users can choose what type of notifier they want Cruise Control to use.
+They can use the notifier provided by Cruise Control, such as the `SelfHealingNotifier`, `SlackSelfHealingNotifier`, `AlertaSelfHealingNotifier` and `MSTeamsSelfHealingNotifier` but then the operator wouldn't be handling the rebalance and Cruise Control will be handling the healing instead, since there is no way for the operator to know whether the anomaly was detected and a rebalance is required ...
```
It's actually not going to happen because all the self.healing parameters are forbidden so the user can't enable self-healing provided from CC.
They can set a notifier but self healing will be anyway disabled. So, no fix in the end?

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+  kind: Kafka
+  metadata:
+    annotations:
+      strimzi.io/anomaly-detected: "GOAL_VIOLATION"
```
Shouldn't we have a little bit more complex structure (using a JSON?) also containing the anomalyId? It would be useful for debugging purposes I guess.

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+  * if there is a queued rebalancing scale down (`Kafka.status.autoRebalance.modes[remove-brokers]` exists), start the rebalancing scale down and transition to **RebalanceOnScaleDown**.
+  * if there is a queued rebalancing scale up (`Kafka.status.autoRebalance.modes[add-brokers]` exists), start the rebalancing scale up and transition to **RebalanceOnScaleUp**.
+  * If no queued rebalancing scale down or scale up, just transition to **Idle**, clean `Kafka.status.autoRebalance.modes`, delete the "actual" `KafkaRebalance` custom resource and remove the annotation .
+* if `PendingProposal`, `ProposalReady` or `Rebalancing`, the rebalancing scale up is still running.
```
Full rebelancing due to anomaly not scale up (I guess just a copy/paste mistake here)

tomncooper reviewed (commented):
I did a first pass. I think this is a better proposal which is more in line with how Strimzi currently uses CC. 

I think you need more detail on the interaction with the current auto-rebalancing and also a clearer description of the FSM states and their transitions. I found it hard to follow the sequence you are proposing. 

For the notifier, I actually think we should stop users using custom notifiers (we could make it conditional on the full mode being set or not). As we are creating K8s resources in response to detected anomalies users can create alerting based on that if they need it. If users do need that then we could provide implementations of the various notifiers which extend our notifier rather than the CC one.

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+The new mode will be called `FULL`, which means that an anomaly was detected and the rebalancing should be applied to the whole cluster
+The mode is defined by setting the `spec.cruiseControl.autoRebalance.mode` field as `full` and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource (read later for more details), by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-defi ...
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+Once the auto-rebalance with `Full` mode is enabled, self-healing will be enabled and a notifier `AnomalyDetectorNotifier` will be set(default notifier).
```
I would also remove the sentence "self-healing will be enabled" or being clear you are referencing our auto-rebalancing triggered by anomalies detection.

ppatierno commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+`SelfHealingNotifier` is the base class that contains the logic of self-healing and override the methods in by implementing the `AnomalyNotifier` interface.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()` etc.
+The `AnomalyDetectorNotifier` will be based of the `SelfHealingNotifier` class.
```
That's a good point because most of the work of the `SelfHealingNotifier` class is dealing with self-healing from CC itself so it has some complex logic inside. I would investigate if just implementing the `AnomalyNotifier` interface would be enough for our implementation.

scholzj commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+* better debugging.
+* to drive the rebalances/fixes from the operator.
+
+#### `Full` mode in auto-rebalancing feature of Strimzi
```
I guess it is `full` mode in lowercase as that is how it will be in the custom resource? (not just here but also in the other places)

scholzj commented on 106-self-healing-feature-in-operator.md:
```diff
@@ -0,0 +1,295 @@
+The anomalies with smaller priority value and detected time will be considered priority and resolved first.
+In case the anomaly is unfixable then they will be ignored.
+The `AnomalyDetectorNotifier` will override the `alert()` method and other detector methods of the `SelfHealingNotifier` to ensure that it is able to alert the operator whenever an anomaly is detected.
+Upon detection of an anomaly, the notifier would apply `strimzi.io/anomaly-detected: <anomaly-type>` annotation on the `Kafka` CR.
```
This is not a good idea as we would run into conflict with whoever or whatever creates the resource. We would need to find other mechanism. For example some simple resource such as ConfigMap or Secret with the violation or have some new internal custom resource for it.

ShubhamRwt replied:
Going though all the three options: configMap, custom resource and rest endpoint, I think the best option would be to use ConfigMaps.

1. We don't have any complex information to show,  the data we need to show would be consisting of the `anomalyId`, `anomalyType`, if possible the timestamp of detection, and the goal that is violated.

3. How does CC clear the violations ... does the CO need to know that it handled given violation?
Good question........So what I understand is that CC would clear violations anyway since we are not using self-healing. The notifier that we will add will only alert by creating a configmap but for the action takes we are probably going to set it as `IGNORE` action which means we want cruise control to ignore the anomaly so no fix is ran. Once the operator has ran the fix we can then delete the configmap associated with that anomaly. The good thing is that in case the rebalance has not fixed the anomaly then that anomaly would be caught again based upon the detection time. But it also raises a question though that if a user configures the detection time to be very short then Cruise Control queue up a lot of Goal violation anomaly which is basically the same anomaly and that can cause an infinite loop. So i think we need to find a way good way to tackle this. @ppatierno what do you think about this situation?

ShubhamRwt replied:
From my previous investigations, when CC used to the do the self healing, it never ran a fix again for an anomaly that was already fixed but not sure how that will behave when we do the rebalance. I will test this case. Thanks @scholzj for finding this issue

scholzj replied:
So, how do you make sure that CC is not raising the same anomaly again and again? E.g. if it things something is out of balance, but the rebalance does not solve this? How do we avoid having a loop of CC triggering the same thing again and again while the CO is trying to solve it again and again without any effect? Is that somehow done by CC to not raise the same thing again for some time?

scholzj replied:
> From my previous investigations, when CC used to the do the self healing, it never ran a fix again for an anomaly that was already fixed but not sure how that will behave when we do the rebalance. I will test this case. Thanks @scholzj for finding this issue

The point is more about the case when it is not fixed. Imagine your disk utilization is out of balance between the brokers. But it is caused by a single huge partition, so running rebalance will not actually solve it, and the _annomaly_ is still there.

scholzj reviewed (commented):
I think this is going in the right direction. But I think it needs to go a bit deeper:
* We need to establish our own terminology and not take over the Cruise Control one. There is not really any _self-healing_ and most of the _anomalies_ are not really anomalies.
* If I read this proposal right, you want to focus on when the cluster is out-of-balance. That is a great start. But perhaps that should not be called `mode: full`? Calling it `full` seems confusing - does it mean that `full` includes scale-up / scale-down? Also, I guess in the future we would add some actual _self-healing_ to handle the broken disks or brokers. That might create additional modes probably. So maybe `mode: rebalance` or `mode: skew` or something like that would make more sense?

ShubhamRwt committed:
Added suggestions by Jakub, Tom and Paolo

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Updated the heading

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Added future scope section

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+# Auto-rebalance on imbalanced clusters
+
+This proposal is about adding support for auto-rebalancing the Kafka cluster in case it gets imbalanced due to some issues like unevenly distributed replicas or overloaded brokers e.t.c.
+When enabled, the Strimzi operator should automatically resolve these issues detected by the Anomaly Detector Manager by running KafkaRebalance via Cruise Control using the KafkaRebalance resource.
```
```suggestion
When enabled, the Strimzi operator should automatically resolve these issues detected by the Anomaly Detector Manager within Cruise Control by using a corresponding KafkaRebalance custom resource.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+This proposal is about adding support for auto-rebalancing the Kafka cluster in case it gets imbalanced due to some issues like unevenly distributed replicas or overloaded brokers e.t.c.
+When enabled, the Strimzi operator should automatically resolve these issues detected by the Anomaly Detector Manager by running KafkaRebalance via Cruise Control using the KafkaRebalance resource.
+Anomalies are detected by Cruise Control using the anomaly detector manager (see section [ Anomaly Detector Manager](./106-auto-rebalance-on-imbalanced-clusters.md#anomaly-detector-manager) below for a detailed description).
```
It's a repetition of the above sentence. Maybe you can delete it but adding the link to the anomaly detector manager to the previous sentence.

see-quick replied:
+1

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+## Motivation
+
+Currently, any anomaly that the user is notified about would need to be fixed manually by using the `KafkaRebalance` custom resource.
```
How a user is notified by anomalies currently? What are you referring to?

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+With smaller clusters, it is feasible to fix things manually. However, for larger ones it can be very time-consuming, or just not feasible, to fix all the anomalies on your own.
+It would be useful for users of Strimzi to be able to have these anomalies fixed automatically whenever they are detected.
+
+### Introduction to Self Healing
```
```suggestion
### Introduction to Self Healing in Cruise Control
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+The above flow diagram depicts the self-healing process in Cruise Control.
+The anomaly detector manager detects an anomaly (using the detector classes) and forwards it to the notifier.
+The configured notifiers provides alerts to the users about the detected anomaly and also returns the action that needs to be taken on the anomaly i.e. whether to fix it, ignore it or delay it.
```
The alert mechanism isn't out of the box. A notifier can have its own logic without generating any alerts. Even just triggering the fix without notifying anyone what's happening. So there is no assumption that a "configured" notifier provides alerts. I think this sentence should say that the notifier makes the decision about the action to take. Then CC provides some notifiers which are able to alert the user in several ways (MS Teams, Slack, etc etc).

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+The detected anomalies can be of various types:
+* Goal Violation - This happens if certain [optimization goals](https://strimzi.io/docs/operators/in-development/deploying#optimization_goals) are violated (e.g. DiskUsageDistributionGoal etc.). These goals can be configured through the `self.healing.goals` option in Cruise Control configuration.  However, this option is forbidden in the `spec.cruiseControl.config` section of the `Kafka` CR.
+* Topic Anomaly - Where one or more topics in cluster violates user-defined properties (e.g. some partitions are too large in disk).
+* Broker Failure - This happens when a non-empty broker crashes or leaves a cluster for time for a long time.
```
```suggestion
* Broker Failure - This happens when a non-empty broker crashes or leaves a cluster for a long time.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+Each detectable anomaly is handled by a specific detector class which then uses another remediation class to run a fix.
+For example, the `GoalViolations` class uses the `RebalanceRunnable` class, the `DiskFailure` class use the `RemoveDisksRunnable` class and so on.
+An optimization proposal (a collection of replica reassignments and partition leadership changes) will then be generated by these `Runnable` classes and that proposal will be applied on the cluster to fix the anomaly.
+In case the anomaly detected is unfixable for e.g. violated hard goals that cannot be fixed typically due to lack of physical hardware (insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy Replica Capacity Goal, or insufficient number of resources to satisfy resource capacity goals), the anomaly wouldn't be fixed and the Cruise Control logs will be upd ...
```
```suggestion
In case the anomaly detected is unfixable for e.g. violated hard goals that cannot be fixed typically due to lack of physical hardware (insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy Replica Capacity Goal, or insufficient number of resources to satisfy resource capacity goals), the anomaly wouldn't be fixed and the Cruise Control will log a warning with the `self-healing is not possible due to unfixable goals` message.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+Users can currently enable anomaly detection and can also [set](https://strimzi.io/docs/operators/latest/full/deploying.html#setting_up_alerts_for_anomaly_detection) the notifier to one of those included with Cruise Control (SelfHealingNotifier, AlertaSelfHealingNotifier, SlackSelfHealingNotifier etc.).
+However, self-healing is disabled and disallowed in Strimzi.
+The `self-healing` properties were disabled in Strimzi's Cruise Control integration because, initially, it was not clear how self-healing would act if pods were rolled in middle of rebalances or how Strimzi triggered manual rebalances should interact with Cruise Control triggered self-healing ones.
```
```suggestion
All the `self.healing` prefixed properties were disabled in Strimzi's Cruise Control integration because, initially, it was not clear how self-healing would act if pods were rolled in middle of rebalances or how Strimzi triggered manual rebalances should interact with Cruise Control triggered self-healing ones.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+They can configure auto-rebalance to enable only for their specific case i.e. setting only `skew` mode or other scaling related modes.
+Once the auto-rebalance with `skew` mode is enabled, the operator will be ready to trigger auto-rebalance whenever the cluster becomes imbalanced.
+To trigger the auto-rebalance, the operator must know that the cluster is imbalanced due to some goal violation anomaly. 
+We will create our own custom notifier named `AnomalyDetectorNotifier` to do the same.
```
We don't know the future but we know you can configure just one notifier within Cruise Control so what about using a more generic name like `StrimziCruiseControlNotifier`, `StrimziNotifier` or similar.
Right now the notifier is doing what we are proposing but it could take other actions in the future.

see-quick replied:
Yeah, that would make it more flexible for future changes, so +1 for naming it more generic way...

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+This notifier's job will be to update the operator regarding the goal violations so that the operator can trigger a rebalance(see section [AnomalyDetectorNotifier](./106-auto-rebalance-on-imbalanced-clusters.md#anomalydetectornotifier))
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
+We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approach.
+For example, when dealing with topic related issues, it will require a coordination with topic operator and metrics issues will require coordination with the Kafka apis.
```
```suggestion
For example, when dealing with topic related issues, it will require a coordination with topic operator and metrics issues will require coordination with the Kafka API.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
+The user can specify rebalancing goals and other configuration for rebalancing, within the resource.
+
+Here is an example template:
```
```suggestion
Here is an example template:

```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,471 @@
+  # ... other rebalancing related configuration
+```
+When the "template" is set the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource based on the "template" when an anomaly is detected and notified by the `AnomalyDetectorNotifier`
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one and also adds the `spec.mode` to it.
````
"adds the `spec.mode` to it." ... with auto-rebalancing it was just a copy of the mode field.
Now it's not a copy, because "skew" doesn't mean anything to a `KafkaRebalance` custom resource but it has to be "full". I would mention that the "skew" mode will be mapped to the "full" mode into the actual `KafkaRebalance` resource.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,471 @@
+```
+When the "template" is set the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource based on the "template" when an anomaly is detected and notified by the `AnomalyDetectorNotifier`
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one and also adds the `spec.mode` to it.
+The generated `KafkaRebalance` custom resource will be called `<my-cluster-name>-auto-rebalancing-skew-anomalyId`, where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, and `skew` refers to applying the rebalance to all the brokers and the anomalyId would be retrieved from the notifier.
````
```suggestion
The generated `KafkaRebalance` custom resource will be called `<my-cluster-name>-auto-rebalancing-skew-<anomalyId>`, where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, and `skew` refers to applying the rebalance to all the brokers and the <anomalyId> would be retrieved from the notifier.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+  finalizers:
+    - strimzi.io/auto-rebalancing
+spec:
+  mode: skew
```
We are using "full" here which means a rebalancing which takes into account all brokers in the cluster (vs add-brokers and remove-brokers which take into account you are adding/removing specific brokers).

im-konge replied:
So it will be always `full` and there will be no new mode like `imbalance` or `skew`, right?

ppatierno replied:
the `imbalance` term is the higher level definition of the mode for the "auto-rebalancing" alongside `add-brokers`, `remove-brokers`. But when a `KafkaRebalance` custom resource is created, the mode will be `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.
To make it clearer, we could have done the same in the past by having the autorebalancing mode definitions of `scalingup` and `scalingdown` then mapped to the `add-brokers` and `remove-brokers` at `KafkaRebalance` level but we went using the same naming instead.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()` etc.
+The `AnomalyDetectorNotifier` will be based of the `AnomalyNotifier` class.
```
```suggestion
The `AnomalyDetectorNotifier` will implement the `AnomalyNotifier` interface.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+The `AnomalyDetectorNotifier` will be based of the `AnomalyNotifier` class.
+The anomalies with smaller priority value and detected time will be considered priority and resolved first.
+In case the anomaly is unfixable due to issues like  lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy ReplicaCapacityGoal, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
+The `AnomalyDetectorNotifier` will override all the abstract methods provided by the `AnomalyNotifier` interface.
```
```suggestion
The `AnomalyDetectorNotifier` will override all the methods declared by the `AnomalyNotifier` interface.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+In case the anomaly is unfixable due to issues like  lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy ReplicaCapacityGoal, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
+The `AnomalyDetectorNotifier` will override all the abstract methods provided by the `AnomalyNotifier` interface.
+We will add an `alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create a configmap with name set as `goal-violation` followed by `anomalyId`.
```
```suggestion
Upon detection of an anomaly, the notifier would create a ConfigMap named as `goal-violation-<anomalyID>`.
```

tomncooper replied:
```suggestion
Upon detection of an anomaly, the notifier will create a `ConfigMap` called `goal-violation-<anomalyId>`.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,471 @@
+# ...
+```
+
+The operator will then check if any configmap with prefix `goal-violation` is created or not, if it finds one created then operator will trigger the rebalance.
````
```suggestion
The operator will then check if any ConfigMap with prefix `goal-violation` is created or not, if it finds one created then operator will trigger the rebalance.
```

tomncooper replied:
I know it is not recommended but what if a user deploys multiple Kafka clusters in the same namespace? How does the operator know what kafka cluster this CM relates too?

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,471 @@
+```
+
+The operator will then check if any configmap with prefix `goal-violation` is created or not, if it finds one created then operator will trigger the rebalance.
+Separate configmaps would be created for every goal violation such that on completion of the rebalance we can remove the particular configmap.
````
```suggestion
Separate ConfigMaps would be created for every goal violation such that on completion of the rebalance we can remove the particular ConfigMap.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+The operator will then check if any configmap with prefix `goal-violation` is created or not, if it finds one created then operator will trigger the rebalance.
+Separate configmaps would be created for every goal violation such that on completion of the rebalance we can remove the particular configmap.
+The advantages of using a separate configmap for every anomaly are:
```
I will not repeat myself anymore :-) "configmap" -> "ConfigMap"

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+Users cannot configure the notifier if they are utilising the auto-rebalance on imbalanced cluster.
+This is because the operator is using our custom notifier for getting alerts about goal violations. 
+If the users try to override the notifier while the `skew` mode is enabled, the auto-rebalance `skew` configuration then the operator would throw errors in the auto-rebalance status field
```
what does it means through errors? Is there a specific state the auto-rebalancing FSM is moving? Or can you provide a yaml snippet with an example about that?

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+If the users really want to have their own way of dealing with the imbalanced clusters then they can just disable auto-rebalance in `skew` mode and use their own notifier.
+
+#### What happens if some unfixable goal violation happens
```
```suggestion
#### What happens if some unfixable goal violation happens

```

tomncooper replied:
```suggestion
#### What happens if an unfixable goal violation happens

```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+We will add an `alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create a configmap with name set as `goal-violation` followed by `anomalyId`.
+
+The ConfigMap would look like this:
```
```suggestion
The ConfigMap will look like this:
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+Users cannot configure the notifier if they are utilising the auto-rebalance on imbalanced cluster.
+This is because the operator is using our custom notifier for getting alerts about goal violations. 
+If the users try to override the notifier while the `skew` mode is enabled, the auto-rebalance `skew` configuration then the operator would throw errors in the auto-rebalance status field
```
Can't we just block the relevant config?

ppatierno replied:
How can we block? The user is going to set their own notifier via the `anomaly.notifier.class` property within the `cruiseControl.config` section and at the same time they enable `skew` mode in the `autoRebalance` section. 
We cannot make `anomaly.notifier.class` forbidden even because we allow the users to use their notifier. It's just it can't be used together with `skew` because the CO is going to use the Strimzi one.

tomncooper replied:
Well, we can check right, you can't use both configs together. If users do then we should log an error somewhere appropriate.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+This is because the operator is using our custom notifier for getting alerts about goal violations. 
+If the users try to override the notifier while the `skew` mode is enabled, the auto-rebalance `skew` configuration then the operator would throw errors in the auto-rebalance status field
+
+If the users really want to have their own way of dealing with the imbalanced clusters then they can just disable auto-rebalance in `skew` mode and use their own notifier.
```
Could the user just extend our notifier? So long as they call `super.alert` first in their `alert` implementation they could add their own notifer logic? Not easy, but it would work.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+If the users really want to have their own way of dealing with the imbalanced clusters then they can just disable auto-rebalance in `skew` mode and use their own notifier.
+
+#### What happens if some unfixable goal violation happens
+In case, there is an unfixable goal violation then the notifier would simply ignore that anomaly and prompt the user about the unfixable violation in the auto-rebalancing status section.
```
```suggestion
If there is an unfixable goal violation, then the notifier would simply ignore that anomaly and prompt the user about the unfixable violation in the auto-rebalancing status section.
```

tomncooper replied:
Would this appear in the Kafka CR? I agree with Paolo that you need an example here.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+* **RebalanceOnScaleUp**: a rebalancing related to a scale up operation is running.
+
+With the new `skew` mode, we will be introducing a new state to the FSM called `RebalanceOnAnomalyDetection`. 
+This state will be related to rebalances related to imbalanced cluster.
```
```suggestion
This state will be associated with rebalances triggered by imbalanced cluster.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+* from **RebalanceOnScaleDown** to:
+  * **RebalanceOnScaleDown**: if a rebalancing on scale down is still running or another one was requested while the first one ended.
+  * **RebalanceOnScaleUp**: if a scale down operation was requested together with a scale up and, because they run sequentially, the rebalance on scale down had the precedence, was executed first and completed successfully. We can now move on with rebalancing for the scale up.
+  * **RebalanceOnAnomalyDetection**: if a configmap related to goal violation was detected. It will run once the queued scale down and scale up is completed
```
I think this ties into your question above Paolo, what happens if a load of Anaomay CMs stack up while you are waiting for scale up or scale down rebalance to finish? 

Even if only on anomaly is detected and a CM created, it could be hours old by the time the scaling operation and rebalance is done. The add/remove-broker rebalances can apply goal fixes as well so they may well fix the original anomaly.

I think you need the concept of freshness for an anomaly. You could just blanket reject (delete) any anomalies detected during an ongoing rebalance.

scholzj replied:
I do not think I can really comment as I do not know how this really works in CC. I raised similar point before with regards to imbalance that cannot be fixed (e.g. because one partition causing the imbalance is too big etc.). Will it be raised again and again? Do we need to somehow detedt those and ignore them? Etc. So this is a bit similar. How do you know it was already resolved or not and will it be repeated or not. 🤷

ShubhamRwt replied:
@tomncooper @ppatierno you are correct, we shuld ignore and delete the  configmap at the same time. I think I didn't mentioned it here which my mistake but later in the flowchart, I if anomalies are detected during a rebalance is happening, we will just ignore that configmap and delete it. As for unfixable anomalies which can keep appearing, there is code present in the Cruise Control `SelfHealingNotifier` which I am going to utilize. That method checks if the rebalance can be performed on the Goal violation or not. If the goal vioaltion cannot be fixed, then we just ognore the anomaly and no configmap would be created in that case

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,471 @@
+  RebalanceOnAnomalyDetection->>Idle: scale up complete with no queued items
+```
+* from **RebalanceOnAnomalyDetection**:
+  * **RebalanceOnAnomalyDetection**: if another goal violation was detected while the first one ended. If a scale down and scale up is also queued up then they will execute first.
````
I don't get what this means? Where are things "queued up"? Where is this queue?

tomncooper replied:
Also your bottom point here says that if a scale down is requested then it will wait for the skew to finish. That seems to contradict what you say here? Or do you mean if a _new_ ROAD is posted that will go after a scale up/down?

ShubhamRwt replied:
Yes, I meant if a new ROAD is posted then that will go after a scale up/down?

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+* if `Ready`, the rebalance was successful.
+  * if there is a queued rebalancing scale down (`Kafka.status.autoRebalance.modes[remove-brokers]` exists), start the rebalancing scale down and transition to **RebalanceOnScaleDown**.
+  * if there is a queued rebalancing scale up (`Kafka.status.autoRebalance.modes[add-brokers]` exists), start the rebalancing scale up and transition to **RebalanceOnScaleUp**.
+  * If no queued rebalancing scale down or scale up, just transition to **Idle**, clean `Kafka.status.autoRebalance.modes`, delete the "actual" `KafkaRebalance` custom resource and also the configmap that triggered the rebalance by matching the `anomalyId` suffix in their names.
```
I think it is still important to be able to see when auto-rebalancing was triggered, for debugging purposes. Does Strimzi emit metrics for when it triggers auto-rebalancing and in what mode?

tomncooper reviewed (commented):
Ok I did another pass. I have a few questions:
- How are you going to distinguish anomaly CMs from different Kafka clusters in the same namespace. I know it is not recommened, but user do deploy multiple Kafka clusters in the same NS. 
- You need to deal with GC'ing all these anomaly CMs in the case where a rebalance is on going. Do you delete them? Do you have some kind of timeout based on the detection interval?
- It is not clear what you mean by scale up/down auto-rebalances being queued up? I assume you mean generated `KafkaRebalance` CRs? But it is not clear.

im-konge commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+# Auto-rebalance on imbalanced clusters
+
+This proposal is about adding support for auto-rebalancing the Kafka cluster in case it gets imbalanced due to some issues like unevenly distributed replicas or overloaded brokers e.t.c.
```
```suggestion
This proposal is about adding support for auto-rebalancing the Kafka cluster in case it gets imbalanced due to some issues like unevenly distributed replicas or overloaded brokers etc.
```

im-konge commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+
+## Future Scope
+In the future, we plan to introduce auto-rebalance for topic and metrics related imbalances.
+AS this feature evolves we can even think of having ways to fix issue like disk failure and broker failures automatically since the fix would be driven by the operator.
```
```suggestion
As this feature evolves we can even think of having ways to fix issues like disk failure and broker failures automatically since the fix would be driven by the operator.
```

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+* if `PendingProposal`, `ProposalReady` or `Rebalancing`, the rebalancing is still running.
+  * No further actions required.
+* if `NotReady`
+  * the rebalancing failed, transition to **Idle** and also removing the corresponding mode from the status. The operator also deletes the "actual" `KafkaRebalance` custom resource and the configmap associated with it.
```
We had the same behaviour even when we use auto rebalance for scaling

ppatierno commented:
@nickgarvey Usually you are able to stop the current rebalancing by applying the `stop` annotation on the `KafkaRebalance` (of course the current batch has to finish first). With auto-rebalancing, the `KafkaRebalance` is owned by the operator and not by the user. That's anyway a good feedback because there is no clear way for the user to stop an auto-rebalancing in progress. I think you could apply the `stop` annotation on the `KafkaRebalance` resource but you can't delete it due to a finalizer. Then you should delete the corresponding mode within the `spec.cruiseControl.autoRebalance.mode` field to avoid the re-triggering. It's something to think about.

see-quick commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+* remove-brokers - auto-rebalancing on scale down
+
+To leverage the automated rebalance on imbalanced cluster, we will be introducing a new mode to the auto-rebalancing feature.
+The new mode will be called `skew`, which means that cluster imbalance was detected and rebalancing should be applied to the all the brokers.
```
Yeah, `imbalance` sounds better in this case.

see-quick commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+The auto-rebalance configuration for the `spec.cruiseControl.autoRebalance.template` property in the `Kafka` custom resource is provided through a `KafkaRebalance` custom resource defined as a "template".
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set.
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
```
```suggestion
This is not an actual rebalance request to get an optimization proposal; it is simply where the configuration for auto-rebalancing is defined.
```

see-quick commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set.
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
+The user can specify rebalancing goals and other configuration for rebalancing, within the resource.
```
```suggestion
The user can specify rebalancing goals and configuration in the resource.
```

Frawless commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+If the users really want to have their own way of dealing with the imbalanced clusters then they can just disable auto-rebalance in `skew` mode and use their own notifier.
+
+#### What happens if some unfixable goal violation happens
+In case, there is an unfixable goal violation then the notifier would simply ignore that anomaly and prompt the user about the unfixable violation in the auto-rebalancing status section.
```
Having prometheus metrics for such cases might be reasonable default way?

Frawless commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,471 @@
+  A[KafkaClusterCreator] --creates--> B[KafkaCluster]
+  B -- calls --> D[KafkaAutoRebalancingReconciler.reconcile]
+  D -- check for configmap with goal-violation prefix --> E{if config map present?}
+  D -- if rebalance in progress --> F[ignore new configmaps and delete them]
```
which CMs will be deleted in that case? My understanding is that `KafkaAutoRebalancingReconciler` will not create new ones. What about these CMs that are used for current rebalancing? These should be deleted by rebalancing itself or?

ShubhamRwt committed:
Added suggestions made on the proposal

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

kyguy commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,488 @@
+To trigger the auto-rebalance, the operator must know that the cluster is imbalanced due to some goal violation anomaly. 
+We will create our own custom notifier named `StrimziCruiseControlNotifier` to do the same.
+This notifier's job will be to update the operator regarding the goal violations so that the operator can trigger a rebalance (see section [AnomalyDetectorNotifier](./106-auto-rebalance-on-imbalanced-clusters.md#anomalydetectornotifier)).
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
```
I am not sure I understand, does this mean that the operator will only trigger a partition rebalance for goal violations that don't require manual intervention?

kyguy commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,488 @@
+We will create our own custom notifier named `StrimziCruiseControlNotifier` to do the same.
+This notifier's job will be to update the operator regarding the goal violations so that the operator can trigger a rebalance (see section [AnomalyDetectorNotifier](./106-auto-rebalance-on-imbalanced-clusters.md#anomalydetectornotifier)).
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
+We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approach.
```
It might be worth coming up with some terminology to distinguish the different types of goal violations that are flagged by the anomaly detector. Then it'd be easier to describe what will and will not be implemented as part of this feature versus features in the future. e.g something like

- **resource distribution violations** : Violations that can be resolved by a partition rebalance. 
- **component failure violations** : Violations that can be resolved through manual intervention (disk or broker failure)
-  ...

We would just need to look at the set of possible violations and divide them into groups based on how they are resolved.

ppatierno replied:
I guess that by `DiskDistributionGoal` you are referring to `DiskUsageDistributionGoal`. Right?
Now you had a valid point regarding non fixable violation. I would assume that when anomaly detector detects a `DiskUsageDistributionGoal` it doesn't know if it's fixable or not yet. Our notifier just adds the goal violation anomaly to the `ConfigMap` and the operator will take care by creating a `KafkaRebalance` resource to fix it. I think that, if the violation is not fixable via a rebalancing, CC will respond by NOT providing a proposal, is that right? Or it will provide a proposal and then when the actual rebalancing is requested then it will fail?

Finally, I agree that we can replace "With this proposal, we are only going to support auto-rebalance on imbalanced cluster." with "With this proposal, we are only going to support auto-rebalance whenever goal violations are detected and the violation can be addressed by a rebalance"

kyguy replied:
> I guess that by DiskDistributionGoal you are referring to DiskUsageDistributionGoal. Right?

Sorry, yes that is what I meant.

>  I think that, if the violation is not fixable via a rebalancing, CC will respond by NOT providing a proposal, is that right? Or it will provide a proposal and then when the actual rebalancing is requested then it will fail?

I believe the former.

> Our notifier just adds the goal violation anomaly to the ConfigMap and the operator will take care by creating a KafkaRebalance resource to fix it.

That makes sense, I guess there will be times an imbalance due to a disk failure could be solved by a rebalance and some times not, we delegate the decision to Cruise Control.

> Finally, I agree that we can replace "With this proposal, we are only going to support auto-rebalance on imbalanced cluster." with "With this proposal, we are only going to support auto-rebalance whenever goal violations are detected and the violation can be addressed by a rebalance"

Sounds good to me!

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,488 @@
+To resolve this issue, we will only make use of Cruise Control's anomaly detection ability, the triggering of the partition reassignments (rebalance) will the responsibility of the Strimzi Cluster Operator.
+To enable this, we will use approach based on the existing auto-rebalance for scaling feature (see the [documentation](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) for more details).
+We will be using the goal violation anomaly detection related classes in Cruise Control to detect imbalanced cluster and not other detection related class like Disk failures or broker failure.
+THe reason behind it is that disk failures and broker failures can be fixed in a much better way than rebalancing the cluster. It is much easier to spin up a new disk in case of disk failures and in the same way it is better to fix the issue with the broker directly instead just moving the partitions replicas away from it.
```
I think the reason I mentioned was because Cruise Control can only do rebalancing and that wouldn't help us fix these failures so therefore we are not supporting it but I think it would be good to frame it the way you mentioned

ShubhamRwt committed:
Refine the proposal

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Pushed example yamls

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Minor Edits

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

kyguy commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+To resolve this issue, we will only make use of Cruise Control's anomaly detection ability, the triggering of the partition reassignments (rebalance) will the responsibility of the Strimzi Cluster Operator.
+To enable this, we will use approach based on the existing auto-rebalance for scaling feature (see the [documentation](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) for more details).
+We will be using the anomaly detection classes related to goal violations that can be addressed by a partition rebalances but not other anomaly detection classes related to goal violations that would require manual intervention like disk or broker failures.
+TThe reason behind thus is that disk failures and broker failures can cannot be fixed by rebalancing alone, they require manual intervention.
```
```suggestion
The reason behind thus is that disk failures and broker failures can cannot be fixed by rebalancing alone, they require manual intervention.
```

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+
+### Auto-rebalancing execution for `imbalance` mode
+
+### Auto-rebalancing Finite State Machine (FSM) for `imbalance` mode
```
```suggestion
#### Auto-rebalancing Finite State Machine (FSM) for `imbalance` mode
```

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+* **RebalanceOnScaleDown**: a rebalancing related to a scale down operation is running.
+* **RebalanceOnScaleUp**: a rebalancing related to a scale up operation is running.
+
+With the new `imbalance` mode, we will be introducing a new state to the FSM called `RebalanceOnAnomalyDetection`.
```
Should it be `RebalanceOnImbalance` instead if the type is imbalance?

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+If, during an ongoing auto-rebalancing, the `KafkaRebalance` custom resource is not there anymore on the next reconciliation, it could mean the user deleted it while the operator was stopped/crashed/not running.
+In this case, the FSM will assume it as `NotReady` so falling in the last case above.
+
+## Affected/not affected projects
```
Should have backwards compatibility section as well to clarify/summarize all the compatibilit issues (the custom notifier I guess being the only one).

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+
+## Affected/not affected projects
+
+This change will affect the Strimzi cluster operator and a new repository named `strimzi-notifier` will be added under the Strimzi organisation.
```
+1 of separate repository for the notifier. But that should be likely already detailed in earlier in the proposal.

scholzj reviewed (commented):
I left some comments. Some are nits, some are questions, etc. I feel like it would be great to have more clarifications on:
* How are the anomalies removed from the CM
* How exactly do we prevent repeated imbalances whcih cannot be fixed.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+Various detector classes like `GoalViolationDetector`, `DiskFailureDetector`, `KafkaBrokerFailureDetector` etc. are used for the anomaly detection, which runs periodically to check if the cluster has their corresponding anomalies or not.
+The frequency of this check can be changed via the `anomaly.detection.interval.ms` configuration.
+Detector classes have different mechanisms to detect their corresponding anomalies.
+For example, `KafkaBrokerFailureDetector` utilises Kafka Metadata API whereas `DiskFailureDetector` and `TopicAnomalyDetector` utilises Kafka Admin API.
```
Yeah, what is that? AFAICS it instantiates a Kafka Admin API based client https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/KafkaBrokerFailureDetector.java#L30

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+* `CHECK` - Delay the anomaly fix
+* `IGNORE` - Ignore the anomaly fix
+
+The default `NoopNotifer` always sets the notifier action as `IGNORE`, which  means that the detected anomaly will be silently ignored and no notification is sent to the user.
```
This sentence can be confusing.
The fact that the notifier returns `IGNORE` doesn't mean that no notification is sent to the user. At least this is how I read the sentence.
We should make it clearer, stating that:

* this notifier returns `IGNORE` which means that the anomaly detector manager won't use any class to fix the anomaly.
* this notifier doesn't implement any notification mechanism to the users

I think it's important because a notifier could return `IGNORE` but at the same time notifying the user, which is NOT what the `NoopNotifer` does.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+
+The default `NoopNotifer` always sets the notifier action as `IGNORE`, which  means that the detected anomaly will be silently ignored and no notification is sent to the user.
+
+Cruise Control also provides [custom notifiers](https://github.com/linkedin/cruise-control/wiki/Configure-notifications) like Slack Notifier, Alerta Notifier etc. for notifying users regarding the anomalies. There are multiple other [self-healing notifier](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations) related configurations you can use to make  ...
```
It's related to the comment I made on line 57.
There is a different between what the notifier says to the anomaly manager, so to fix or not the anomaly, and what the notifier notifies to the user.
The FIX, CHECK and IGNORE values are for the anomaly manager to understand if the anomaly should be fixed or not.
Then the notifier itself can send notification to the user or interact with the user in general in a different way.
The Slack notifier relies on the base notifier implementation to make the decision if the anomaly should be fixed or not. At the same time then it sends a message on Slack to the user. So to answer your question it doesn't always return IGNORE but always sends a Slack message with details about the anomaly and the action taken.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+Even under normal operation, it's common for Kafka clusters to encounter problems such as partition key skew leading to an uneven partition distribution, or hardware issues like disk failures, which can degrade overall cluster's health and performance.
+Currently, in any such scenario these issues need to be fixed manually i.e. if the cluster is imbalanced then a user might instruct Cruise Control to move the partition replicas across the brokers in order to fix the imbalance using the `KafkaRebalance` custom resource.
+
+Users can currently enable anomaly detection and can also [set](https://strimzi.io/docs/operators/latest/full/deploying.html#setting_up_alerts_for_anomaly_detection) the notifier to one of those included with Cruise Control (`SelfHealingNotifier`, `AlertaSelfHealingNotifier`, `SlackSelfHealingNotifier` etc.).
```
There is no limitation today. You are right the users can also set their own notifier, to get notification the way they prefer. At the same time, today, they can't enable the self-healing (see below, `self.healing` fields are forbidden). So, long story short, today the users can leverage the anomaly detection and notifier but not using the self-healing (auto-fix anomalies).

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+
+This proposal allows the users to have their cluster balanced automatically whenever the cluster gets imbalanced due to overloaded broker, CPU usage etc.
+If we were to enable the self-healing ability of Cruise Control then, in response to detected anomalies, Cruise Control would issue partition reassignments without involving the Strimzi Cluster Operator.
+This could cause potential conflicts with other administration operations and is the primary reason self-healing has been disabled until now.
```
I would leave it here, but at the same time agree with Kyle to put the "self-healing provided by CC" as a rejected alternatives for the various reasons.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+* remove-brokers - auto-rebalancing on scale down
+
+To leverage the automated rebalance on imbalanced cluster (those with detected Goal violations), we will be introducing a new mode to the auto-rebalancing feature.
+The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to the all the brokers.
```
```suggestion
The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to all the brokers.
```

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+This notifier's job will be to update the operator regarding the goal violations so that the operator can trigger a rebalance (see section [AnomalyDetectorNotifier](./106-auto-rebalance-on-imbalanced-clusters.md#anomalydetectornotifier)).
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
+We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approach.
+For example, when dealing with topic related issues, it will require a coordination with topic operator and metrics issues will require coordination with the Kafka API.
```
I don't know the internal details but CC provides several topic related "fixers" for anomalies related to topic which could involve changing the replication factor for example. But AFAIU, it would be CC using the Kafka Admin API to do so, while we have our TO doing that with `KafkaTopic` being the source of truth.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+
+The auto-rebalance configuration for the `spec.cruiseControl.autoRebalance.template` property in the `Kafka` custom resource is provided through a `KafkaRebalance` custom resource defined as a "template".
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set.
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
```
I would just mention the Strimzi Cluster Operator than the specific assembly operator class.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()`.
+The `AnomalyDetectorNotifier` will implement the `AnomalyNotifier` interface.
```
Yeah I think it should be `StrimziCruiseControlNotifier` here and even in some of the following text.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+data:
+  last_anomaly_detection_timestamp: <when last anomaly detection was run>
+  rebalance_finished_timestamp: <when the rebalance finished>
+  anomaly_list: |
```
Why not using camel case for this naming instead of the `_` separator?

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,524 @@
+```
+
+If the users really want to have their own way of dealing with the imbalanced clusters then they can disable auto-rebalance in `imbalance` mode and use their own notifier. 
+Another way for users to use their own notifier can be to extend our notifier and use our alert method i.e `super.alert()` first in their `alert()` method implementation.
````
Jakub is right. They can use our notifier as a base but then it won't be used together with our "imbalance" auto-rebalancing mechanism. So maybe it doesn't make much sense advice they can extend our our notifier.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+
+This state is set since the beginning when a `Kafka` custom resource is created with the `spec.cruiseControl.autoRebalance` field.
+It is also the end state of a previous successfully completed or failed auto-rebalancing.
+In case of successful completion, once the rebalance moves to `Ready` state, we will delete the KafkaRebalance and move the empty the `anomaly_list` then update the `auto-rebalance` state to `Idle`.
```
```suggestion
In case of successful completion, once the rebalance moves to `Ready` state, we will delete the `KafkaRebalance` and move the empty the `anomaly_list` then update the `auto-rebalance` state to `Idle`.
```

ShubhamRwt committed:
Added suggestions

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+Various detector classes like `GoalViolationDetector`, `DiskFailureDetector`, `KafkaBrokerFailureDetector` etc. are used for the anomaly detection, which runs periodically to check if the cluster has their corresponding anomalies or not.
+The frequency of this check can be changed via the `anomaly.detection.interval.ms` configuration.
+Detector classes use different mechanisms to detect their corresponding anomalies.
+For example, `KafkaBrokerFailureDetector` utilises Kafka Admin API whereas `DiskFailureDetector` and `TopicAnomalyDetector` utilises Kafka Admin API.
```
So they both use the Kafka Admin API? How is that a different mechanism?

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+The frequency of this check can be changed via the `anomaly.detection.interval.ms` configuration.
+Detector classes use different mechanisms to detect their corresponding anomalies.
+For example, `KafkaBrokerFailureDetector` utilises Kafka Admin API whereas `DiskFailureDetector` and `TopicAnomalyDetector` utilises Kafka Admin API.
+Furthermore, `MetricAnomalyDetector` use metrics and `GoalViolationDetector` uses the load distribution to detect their anomalies.
```
```suggestion
Furthermore, `MetricAnomalyDetector` use metrics and `GoalViolationDetector` uses the load distribution (calculated by Cruise Control itself) to detect their anomalies.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+For example, `KafkaBrokerFailureDetector` utilises Kafka Admin API whereas `DiskFailureDetector` and `TopicAnomalyDetector` utilises Kafka Admin API.
+Furthermore, `MetricAnomalyDetector` use metrics and `GoalViolationDetector` uses the load distribution to detect their anomalies.
+The detected anomalies can be of various types:
+* Goal Violation - This happens if certain [optimization goals](https://strimzi.io/docs/operators/in-development/deploying#optimization_goals) are violated (e.g. DiskUsageDistributionGoal etc.). These goals can be configured through the `self.healing.goals` option in Cruise Control configuration.  However, this option is forbidden in the `spec.cruiseControl.config` section of the `Kafka` CR.
```
```suggestion
* Goal Violation - This happens if certain [optimization goals](https://strimzi.io/docs/operators/in-development/deploying#optimization_goals) are violated (e.g. DiskUsageDistributionGoal etc.). These goals can be configured through the `self.healing.goals` option in the Cruise Control configuration.  However, this option is currently forbidden in the `spec.cruiseControl.config` section of the `Kafka` CR.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+
+### Proposal
+
+This proposal allows the users to have their cluster balanced automatically whenever the cluster gets imbalanced due to overloaded broker, CPU usage etc.
```
```suggestion
This proposal is concerned with giving users the option to have their Kafka cluster's balanced automatically whenever they become imbalanced due to overloaded broker, CPU usage etc.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+If we were to enable the self-healing ability of Cruise Control then, in response to detected anomalies, Cruise Control would issue partition reassignments without involving the Strimzi Cluster Operator.
+This could cause potential conflicts with other administration operations and is the primary reason self-healing has been disabled until now.
+To resolve this issue, we will only make use of Cruise Control's anomaly detection ability, the triggering of the partition reassignments (rebalance) will the responsibility of the Strimzi Cluster Operator.
+To enable this, we will use approach based on the existing auto-rebalance for scaling feature (see the [documentation](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) for more details).
```
```suggestion
To enable this, we will use an approach based on the existing auto-rebalance for scaling feature (see the [documentation](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) for more details).
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+This could cause potential conflicts with other administration operations and is the primary reason self-healing has been disabled until now.
+To resolve this issue, we will only make use of Cruise Control's anomaly detection ability, the triggering of the partition reassignments (rebalance) will the responsibility of the Strimzi Cluster Operator.
+To enable this, we will use approach based on the existing auto-rebalance for scaling feature (see the [documentation](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) for more details).
+We will be using the anomaly detection classes related to goal violations that can be addressed by a partition rebalances but not other anomaly detection classes related to goal violations that would require manual intervention like disk or broker failures.
```
```suggestion
We propose using only those anomaly detection classes, related to goal violations, that can be addressed by a partition rebalance. 
We will not enable the other anomaly detection classes, related to goal violations, that would require manual interventions at the infrastructure level such as disk or broker failures.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+To enable this, we will use approach based on the existing auto-rebalance for scaling feature (see the [documentation](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) for more details).
+We will be using the anomaly detection classes related to goal violations that can be addressed by a partition rebalances but not other anomaly detection classes related to goal violations that would require manual intervention like disk or broker failures.
+The reason behind this is that disk failures and broker failures can cannot be fixed by rebalancing alone, they require manual intervention.
+It would much easier to spin up a new disk in case of disk failures and in the same way it is better to fix the issue with the broker directly instead of just moving the partitions replicas away from it through rebalancing. 
```
```suggestion
For the latter case, it would be better to spin up new disks or to fix the issue with the broker(s) directly instead of just moving the partitions replicas away from them through rebalancing. 
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+We will be using the anomaly detection classes related to goal violations that can be addressed by a partition rebalances but not other anomaly detection classes related to goal violations that would require manual intervention like disk or broker failures.
+The reason behind this is that disk failures and broker failures can cannot be fixed by rebalancing alone, they require manual intervention.
+It would much easier to spin up a new disk in case of disk failures and in the same way it is better to fix the issue with the broker directly instead of just moving the partitions replicas away from it through rebalancing. 
+Therefore, it would be non-trivial for the Strimzi Operator to fix these failures, hence these are not goals of this proposal and might be addressed later in a separate proposals.
```
```suggestion
Therefore, given the interventions required, it would be non-trivial for the Strimzi Operator to fix these failures automatically. 
Hence, these are not goals of this proposal and could be addressed later in separate proposals.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+
+To leverage the automated rebalance on imbalanced cluster (those with detected Goal violations), we will be introducing a new mode to the auto-rebalancing feature.
+The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to all the brokers.
+The mode is defined by setting the `spec.cruiseControl.autoRebalance.mode` field as `imbalance` and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-refe ...
```
```suggestion
The imbalance mode is configured by creating a new object in the `spec.cruiseControl.autoRebalance` list with its `mode` field set to `imbalance` and the corresponding rebalancing configuration is defined as a reference to a template `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+To leverage the automated rebalance on imbalanced cluster (those with detected Goal violations), we will be introducing a new mode to the auto-rebalancing feature.
+The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to all the brokers.
+The mode is defined by setting the `spec.cruiseControl.autoRebalance.mode` field as `imbalance` and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-refe ...
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration (i.e. the same used for unmodified manual `KafkaRebalance` invocations).
```
```suggestion
The `template` field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration (i.e. the same used for unmodified manual `KafkaRebalance` invocations).
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set.
+When it is created, the cluster operator doesn't run any rebalancing.
+This is not an actual rebalance request to get an optimization proposal; it is simply where the configuration for auto-rebalancing is defined.
+The user can specify rebalancing goals and configuration in the resource.
```
This sentence is redundant, you have said this above.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,534 @@
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+When the "template" is set the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource based on the "template" when an anomaly is detected and notified by the `StrimziCruiseControlNotifier`
````
Why is template in quotation marks everywhere? Can you remove them.
```suggestion
When the template is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when an anomaly is detected and notified by the `StrimziCruiseControlNotifier`
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+
+The operator also sets a finalizer, named `strimzi.io/auto-rebalancing`, on the "actual" `KafkaRebalance` custom resource.
+This is needed to avoid the user, or any other tooling, to delete the resource while the auto-rebalancing is still running.
+The finalizer is removed when Cruise Control indicates that the partition reassignment (rebalance) process has finished, with or without errors, allowing the "actual" `KafkaRebalance` custom resource deletion by the operator itself.
```
```suggestion
The finalizer is removed when Cruise Control indicates that the partition reassignment (rebalance) process has finished, with or without errors, allowing the generated `KafkaRebalance` custom resource to be deleted by the operator itself.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+The `StrimziCruiseControlNotifier` will implement the `AnomalyNotifier` interface.
+Every detected anomaly is placed in a priority queue and each of the anomalies have a priority value assigned to them. 
+Since we are working with only goal related violation every anomaly would have same priority, and they will be fixed on the basis of first come, first served mechanism.
+In case the anomaly is unfixable due to issues like a lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy `ReplicaCapacityGoal`, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
```
```suggestion
In situations where the anomaly is unfixable, due to issues like a lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy `ReplicaCapacityGoal`, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
```

tomncooper replied:
What happens if CC is firing loads of unfixable anomaly alerts? How would that be surfaced to the user? Metrics? Logs? It would be worth mentioning that here.

tomncooper replied:
I see you have answered this is in the proposal section, you may want to mention that the suer will be alerted via a Condition in the Kafka CR status.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,534 @@
+# ...
+```
+
+The advantages of using a single ConfigMap for every anomaly are:
````
```suggestion
The advantages of using a single ConfigMap for all the detected anomalies versus one per anomaly are:
```
I think that is what you mean?

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+In case the anomaly is unfixable due to issues like a lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy `ReplicaCapacityGoal`, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
+The `StrimziCruiseControlNotifier` will override all the methods declared by the `AnomalyNotifier` interface.
+We will provide a concrete implementations of the `AnomalyNotifier.alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create a ConfigMap with name set as `<cluster-name>-goal-violation-map`. 
```
```suggestion
Upon detection of an anomaly, the notifier would create (if it doesn't already exist) or update a ConfigMap with the name `<cluster-name>-goal-violation-map`. 
This ConfigMap will contain all the detected anomalies for a given Kafka cluster.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+
+The advantages of using a single ConfigMap for every anomaly are:
+1. Single place to put in all the anomalies
+2. Even when dealing multiple Kafka clusters, the ConfigMap created will be equal to number of clusters deployed
```
```suggestion
2. When dealing multiple Kafka clusters, the ConfigMap created will be equal to number of clusters deployed
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+
+The created ConfigMap will persist when self-healing is enabled and will be deleted if the user decides to disable self-healing. This is because the ConfigMap is going to help us track the timestamps during the complete process.
+
+#### What happens if same anomaly is detected while the auto-rebalance is happening
```
```suggestion
#### What happens if same anomaly type is detected while an `imbalance` auto-rebalance is running
```
You won't know if it is exactly the same one right, only that it is the same type?

tomncooper replied:
This section needs to be much clearer. Spell out the potential issue, the various scenarios, how logging and updating the time stamps will work and how that helps fix the issues under each scenario.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,534 @@
+        Notifier cannot be overridden since `imbalance` mode is enabled in the `auto-rebalance` configuration
+```
+
+If the users really want to have their own way of dealing with the imbalanced clusters then they can disable auto-rebalance in `imbalance` mode and use their own notifier. 
````
It may be possible for our notifier to forward notifications to other configured notifiers, but this would a future enhancement if users want it.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+
+#### Metrics for tracking the rebalance requests
+
+If the users want to track when the auto-rebalance happened or not, they can access the Strimzi [metrics](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/grafana-dashboards/strimzi-operators.json#L712) about when the `KafkaRebalance` custom resources were visible/created. These metrics also cover the `KafkaRebalance`(s) which were created automatically so the users ca ...
```
```suggestion
If the users want to track when the auto-rebalances happen, they can access the Strimzi [metrics](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/grafana-dashboards/strimzi-operators.json#L712) for the `KafkaRebalance` custom resources, this includes when they were visible/created.
These metrics also cover the `KafkaRebalance`(s) which were created automatically, so the users can utilize them to understand when an auto-rebalance were triggered in their cluster.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+  A[KafkaClusterCreator] --creates--> B[KafkaCluster]
+  B -- calls --> D[KafkaAutoRebalancingReconciler.reconcile]
+  D -- check for ConfigMap with goal-violation prefix --> E{if config map present?}
+  D -- if rebalance in progress --> F[moves a;; unfixed anomalies in list to fixed anomalies list]
```
What is are these "unfixed" and "fixed anomalies list"(s)? This is the first time you have mentioned them?

tomncooper replied:
```suggestion
  D -- if rebalance in progress --> F[moves unfixed anomalies in list to fixed anomalies list]
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
+2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check if there was any ConfigMap created with name  `goal-violation-map` and whether the `anomaly_list` list is empty or not, then the `full` rebalance(imbalance mode) would be performed if the list is not empty.
+If the list is not empty, then we trigger the `imbalance` mode (full rebalance). 
```
This sentence just repeats the end of the one before?

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,534 @@
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
+2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check if there was any ConfigMap created with name  `goal-violation-map` and whether the `anomaly_list` list is empty or not, then the `full` rebalance(imbalance mode) would be performed if the list is not empty.
+If the list is not empty, then we trigger the `imbalance` mode (full rebalance). 
+3. In case, a rebalance is already ongoing and more anomalies are detected, then the operator will just ignore the new anomalies and delete all the anomalies from the `anomaly_list` in the ConfigMap.
```
```suggestion
3. If a rebalance is already ongoing and more anomalies are detected, then the operator will just ignore the new anomalies and delete all the anomalies from the `anomaly_list` in the ConfigMap.
```

tomncooper reviewed (commented):
I have had another pass, mostly just suggestions around language and grammar.

However, you need a much clearer description of what happens when anomalies are detected when a rebalance is ongoing (either auto or manually triggered). Why is that an issue, what scenarios could arise, how does having the timestamps help avoid that. Also, what exactly happens to the anomaly list, when is it updated, when is it cleared etc.

tomncooper reviewed (commented):
I have had another pass, mostly just suggestions around language and grammar.

However, you need a much clearer description of what happens when anomalies are detected when a rebalance is ongoing (either auto or manually triggered). Why is that an issue, what scenarios could arise, how does having the timestamps help avoid that. Also, what exactly happens to the anomaly list, when is it updated, when is it cleared etc.

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,534 @@
+```
+When the "template" is set the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource based on the "template" when an anomaly is detected and notified by the `StrimziCruiseControlNotifier`
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one.
+The "imbalance" mode will be mapped to the "full" mode in the actual KafkaRebalance resource which means that  actual `KafkaRebalance` custom resource will have the mode set as `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.
````
The intro specifies the 2 modes that are present in `auto-rebalancing` feature. Here I was talking about mapping the new `imbalance` of `auto-rebalancing` feature to the `full` mode of KafkaRebalance resource.

ShubhamRwt committed:
Added suggestions by Tom and Jakub

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Minor fixes

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+The `anomalyList` will be populated by the notifier everytime it detects an anomaly
+Whenever a rebalance is about to start by the operator(either it be in imbalance mode, remove-broker mode, add-broker mode) the `anomalyList` would be set to empty.
+We empty the list earlier to make it easier to track the anomalies that were caught in between the rebalance.
+If the `anomalyList` isn't empty after the rebalance that means that there were anomalies detected in between the rebalance, and we need to make decision on whether we run another rebalance for these detected anomalies or not.
```
```suggestion
If the `anomalyList` isn't empty after a rebalance completes, that means that there were anomalies detected in whilst the rebalance was running.
We then need to make a decision on whether we run another rebalance for these detected anomalies or not.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+To tackle this problem we can make use of the `last_anomaly_detection_timestamp` and `rebalance_finished_timestamp` timestamps
+Everytime an anomaly of the same type is detected we will treat its detection time as the `last_anomaly_detection_time` and update the value of it in the `last_anomaly_detection_time` field of the ConfigMap.
+Now if the anomaly detection happened during the rebalance, the `ConfigMap` will be updated with the anomalies as well as the `last_anomaly_detection_time` will also be updated with last anomaly detection time.
+We will track the rebalance finish time in ConfigMap under the `rebalance_finished_timestamp` field.
```
You have basically already said this (regarding both timestamps) in the section above. I don't think you need to spell it out again.

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+Everytime an anomaly of the same type is detected we will treat its detection time as the `last_anomaly_detection_time` and update the value of it in the `last_anomaly_detection_time` field of the ConfigMap.
+Now if the anomaly detection happened during the rebalance, the `ConfigMap` will be updated with the anomalies as well as the `last_anomaly_detection_time` will also be updated with last anomaly detection time.
+We will track the rebalance finish time in ConfigMap under the `rebalance_finished_timestamp` field.
+Now to know if the new anomaly was detected during the rebalance we can compare the `last_anomaly_detection_timestamp` and `rebalance_finished_timestamp`.
```
```suggestion
To determine if the new anomaly was detected during the rebalance we can compare the `last_anomaly_detection_timestamp` and `rebalance_finished_timestamp`.
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+This is because it is possible that the new anomalies were fixed during the rebalance.
+Even if the ignored anomalies weren't fixed, the Anomaly Detector can detect them again.
+
+The above hold true for the manual rebalances also. 
```
```suggestion
The above also holds true for the manual rebalances. 
```

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+
+Cruise Control has the ability to figure out if a goal violation can be fixed by the generated optimization proposal or not.
+The goal optimizer class does this job by doing analysis/dry-run if we can fix the anomaly by the generated proposal or not.
+If there is some goal, for example the `DiskDistributionUsage` goal is violated, and we cannot fix it since the all the disks are already completely populated then it is marked as unfixable.
```
```suggestion
If a goal such as the `DiskDistributionUsage` goal is violated, but we cannot fix it since the all the disks are already completely populated, then it is marked as unfixable.
```

tomncooper reviewed (commented):
The section on the StrimziCruiseControlNotifier, particularly the CM format needs review. It is not clear what the unfixable anomalies/goal list is actually going to contain? You also need to make clear what happens if all the detected anomalies are unfixable?

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+This generated ConfigMap will contain the list of goal violation anomalies in the section named `anomalyList`.
+The `anomalyList` will refer to the anomalies that were detected by the anomaly detector.
+The `anomalyList` will be populated by the notifier everytime it detects an anomaly
+Whenever a rebalance is about to start by the operator(either it be in imbalance mode, remove-broker mode, add-broker mode) the `anomalyList` would be set to empty.
```
What if someone has autorebalancing on imbalance configured and the notifier creates the CM when it detects an anomaly, but the user posts a manual KafkaRebalance before the reconcilation loop triggers the Operator to look at the anomaly CM? 

Or is the operator able to watch for a specific CM resource, so it can trigger as soon as the anomaly CM is created/updated?

ppatierno replied:
I got what you mean here, I would assume if the manual triggers first then yes the anomalyList will be cleared.

> Or is the operator able to watch for a specific CM resource, so it can trigger as soon as the anomaly CM is created/updated?

Well the operator has a watch for both ... KafkaRebalance and ConfigMap changes ... so which comes first to be notified by Kube API.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,524 @@
+* from **RebalanceOnScaleDown** to:
+  * **RebalanceOnScaleDown**: if a rebalancing on scale down is still running or another one was requested while the first one ended.
+  * **RebalanceOnScaleUp**: if a scale down operation was requested together with a scale up and, because they run sequentially, the rebalance on scale down had the precedence, was executed first and completed successfully. We can now move on with rebalancing for the scale up.
+  * **Idle**: if a scale down operation was requested, it was executed and completed successfully/failed or a full rebalance was asked due to an anomaly but since the scale-down rebalance is done, we can ignore the anomalies assuming they are fixed by the rebalance. In case, they are not fixed, Cruise Control will detect them again and a new rebalance would be requested.
```
I re-opened this because I had the same thought as Tom. The `RebalanceOnImbalance` is in the other diagram below from `RebalanceOnScaleUp` because a ConfigMap is detected so it should be also here after a scale down rebalance.

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,583 @@
+```mermaid
+flowchart TB
+  A[KafkaClusterCreator] --creates--> B[KafkaCluster]
+  B -- calls --> D{Reconcilation in Progress}
````
I think Shubham is really referring to the operator reconciliation, but of course I could be wrong :-)

But I have following comments:
* I think the diagram should not show the `KafkaClusterCreator` and `KafkaCluster` which are part of the general flow and not important in describing the auto-rebalancing.
* I can't map what you say on point 2 with the arrows "no" and "yes" going out of "Reconciliation in Progress" box.
* isn'g "fix the anomalies" the same as "Trigger auto-rebalance" ? 

I am confused sorry, I can't understand the goal of the diagram and how it maps on the description below.

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+If the user runs a `manual` rebalance, we will empty the `anomalyList` in those cases too.
+If some other anomalies are detected again after the rebalance then they will be updated in the ConfigMap by the notifier and another rebalance will take place
+
+The ConfigMap will also have an option field called `unfixedAnomaly` list which would be created only when an unfixable anomaly is detected.
```
The name of the list would be `unfixableGoalList`. I will fix the name. If all the anomalies are unfixable they will be added to the `unfixableGoal` list. Our  code will check the `anomalyList` for triggering the rebalance. So therefore no rebalance will happens if all the detected anomalies are unfixable

tinaselenge commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+* from **Idle** to:
+  * **RebalanceOnScaleDown**: if a scale down operation was requested. This transition happens even if a scale up was requested at the same time but the rebalancing on scaling down has the precedence. The rebalancing on scale up is queued. They will run sequentially.
+  * **RebalanceOnScaleUp**: if only a scale up operation was requested. There was no scale down operation requested.
+  * **RebalanceOnImbalance**: if a ConfigMap related to goal violation was detected, and the `anomalyList` is not empty
```
I think this sentence needs to cover a similar information like the other 2 above in term of which operation has the precedence, if other operations requested at the same time. I understand the diagram tries to explain it but without a little explanation here, it's not easy to understand.

tinaselenge commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnImbalance: then full rebalance
+  end
+  RebalanceOnScaleDown->>Idle: scale down complete with no queued items
```
I find these diagrams quite confusing. I think as @ppatierno suggested, simpler diagrams with states/triggers/arrows would be better.

ShubhamRwt replied:
I am working on this

tinaselenge commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
+2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check if there was any ConfigMap created with name  `goal-violation-map` and whether the `anomalyList` list is empty or not, then the `full` rebalance(imbalance mode) would be performed if the list is not empty.
+3. If a rebalance is already ongoing and more anomalies are detected, then the operator will just ignore the new anomalies based on the timestamps as discussed earlier in the proposal and delete all the anomalies from the `anomalyList` in the ConfigMap.
```
I don't think these description matches the diagram very well. I understand the diagram is high level and then you have low level descriptions here, maybe that makes it confusing.

tinaselenge reviewed (commented):
Thank you for the update Shubham. It's been a while since the last time I read the proposal, and can see quite a lot changed. I like the proposed solution to use the auto-rebalancing and let the operator handle the healing. 

Overall I'm happy with the proposal, but I think there are a few parts that need to be ironed out with more clarity.

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+
+To tackle this problem we can make use of the `last_anomaly_detection_timestamp` and `rebalance_finished_timestamp` timestamps
+Everytime an anomaly of the same type is detected we will treat its detection time as the `last_anomaly_detection_time` and update the value of it in the `last_anomaly_detection_time` field of the ConfigMap.
+Now if the anomaly detection happened during the rebalance, the `ConfigMap` will be updated with the anomalies as well as the `last_anomaly_detection_time` will also be updated with last anomaly detection time.
```
The `last_anomaly_detection_time` would be updated every time an anomaly is detected so that we keep track of the when the last anomaly detection was ran

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,583 @@
+The imbalance mode is configured by creating a new object in the `spec.cruiseControl.autoRebalance` list with its `mode` field set to `imbalance` and the corresponding rebalancing configuration is defined as a reference to a template `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kub ...
+The `template` field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration (i.e. the same used for unmodified manual `KafkaRebalance` invocations).
+To provide users more flexibility, they only have to configure the auto-rebalance modes they wish to use whether it be `add-brokers`, `remove-brokers`, or `imbalance`.
+When the auto-rebalance configuration is set with `imbalance` mode enabled, the operator will trigger a partition rebalance whenever a goal violation is detected by the anomaly detector.
```
I have shared the Kafka CR example after some lines(line 121)

ShubhamRwt committed:
Added suggestions by Tina, Tom and Paolo

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

katheris commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,507 @@
+We will create our own custom notifier named `StrimziCruiseControlNotifier` to do this.
+This notifier's job will be to update the operator regarding the goal violations so that the operator can trigger a rebalance (see section [StrimziCruiseControlNotifier](./106-auto-rebalance-on-imbalanced-clusters.md#strimzicruisecontrolnotifier)).
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
+We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approach.
```
```suggestion
We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approaches.
```

katheris commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,507 @@
+When the `template` is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when an anomaly is detected and notified by the `StrimziCruiseControlNotifier`
+The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
+If the user has not configured the anomaly detection goals in Cruise Control then the operator would set the default goals to be used by the anomaly detector.
+The goals mentioned in the rebalance template will be validated with the anomaly detection goals being used by Cruise Control. The mentioned goals should be either same or be a subset of the anomaly detection goals since the unmentioned goals will not be detected and therefore no rebalance would be required for those goals.
```
Wording suggestions to make this clearer. I'm making some assumptions here, so please validate that I've understood correctly.
```suggestion
If the user specifies a rebalance template then the goals mentioned in the rebalance template will be validated against the anomaly detection goals being used by Cruise Control.
The goals in the rebalance template should be either the same or be a subset of the anomaly detection goals being used by Cruise Control.
This is to ensure rebalances are only being run with goals that Cruise Control is checking for anomalies.
```

ShubhamRwt replied:
Yes, you understood correct

katheris commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,507 @@
+
+This generated ConfigMap will contain the list of fixable goal violation anomalies in the section named `fixableAnomalies`.
+The `fixableAnomalies` list will state the `anomalyId` of anomaly as well as the fixable violated goals.
+The `fixableAnomalies` list will be populated by the notifier everytime it detects an anomaly.
```
```suggestion
The `fixableAnomalies` list will be populated by the notifier every time it detects an anomaly.
```

katheris commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,507 @@
+#### Metrics for tracking the rebalance requests
+
+If the users want to track when the auto-rebalances happen, they can access the Strimzi [metrics](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/grafana-dashboards/strimzi-operators.json#L712) for the `KafkaRebalance` custom resources, this includes when they were visible/created.
+These metrics also cover the `KafkaRebalance`(s) which were created automatically, so the users can utilize them to understand when an auto-rebalance were triggered in their cluster.
```
```suggestion
These metrics also cover the `KafkaRebalance`(s) which were created automatically, so the users can utilize them to understand when an auto-rebalance was triggered in their cluster.
```

katheris reviewed (commented):
Thanks @ShubhamRwt. I've added some formatting suggestions and a couple of clarifying questions.

The only other area I wondered about was whether we should consider making use of the maintenance windows? So for example not triggering a rebalance unless we are within a maintenance window. What do you think @scholzj ?

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,507 @@
+```
+When the `template` is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when an anomaly is detected and notified by the `StrimziCruiseControlNotifier`
+The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
+If the user has not configured the anomaly detection goals in Cruise Control then the operator would set the default goals to be used by the anomaly detector.
````
Yes, I am referring to Cruise Control section of the Kafka CR. If no anomaly detection goals are set by the user, then the operator set these anomaly goals as default ->  RACK_AWARENESS_GOAL, MIN_TOPIC_LEADERS_PER_BROKER_GOAL, REPLICA_CAPACITY_GOAL, DISK_CAPACITY_GOAL

ShubhamRwt committed:
Added suggestions by Kate

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,510 @@
+  * **Idle**: if a scale down operation was requested, it was executed and completed successfully/failed or a full rebalance was asked due to an anomaly but since the scale-down rebalance is done, we can ignore the anomalies assuming they are fixed by the rebalance. In case, they are not fixed, Cruise Control will detect them again and a new rebalance would be requested.
+
+* from **RebalanceOnScaleUp**:
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is still running or another one was requested while the first one ended.
```
to be fixed in the diagram by adding "scale up is still running"

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,510 @@
+* from **RebalanceOnScaleUp**:
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is still running or another one was requested while the first one ended.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, so the current rebalancing scale up is stopped (and queued) and a new rebalancing scale down is started. The rebalancing scale up will be postponed.
+  * **RebalanceOnImbalance**: if a ConfigMap related to goal violation was detected and the `fixableAnomalies` list is not empty while the `unfixableAnomalies` list is empty. It will run once the queued scale down and scale up is completed.
```
The same comment I made in the `from **RebalanceOnScaleDown**` applies here as well.

tinaselenge reviewed (commented):
Overall, it looks really good to me. I left a couple of questions to clarify my understanding. Thanks!

ShubhamRwt committed:
Added suggestions by Paolo and Tina

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Update image

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

katheris reviewed (commented):
Thanks for addressing my previous comments @ShubhamRwt. I added one more related to how the user can clear the unfixable list in the ConfigMap.

I also saw a bunch of discussion about FSM and want to confirm my understanding. So from what I understand basically the `RebalanceOnImbalance` state is only reachable from `Idle`. This is because when a new anomaly is detected if there is a rebalance currently ongoing it might fix the anomaly, so we don't ever queue an imbalance related rebalance, is that right?

Then if an imbalance related rebalance is ongoing and a scale up or scale down is needed we queue those in the same way we do today.

Am I understanding all of that right?

ShubhamRwt commented:
Yes @katheris you are right for the case where we are reachable to `RebalanceOnImbalance` only from Idle state.  In case, if an imbalance related rebalance is ongoing, and a scale up or scale down is asked we will stop the imbalance.(just like we do today for scale up in case a scale down is asked). So if a  `RebalanceonImbalance` is ongoing and any scale up operation happens, we will stop the imbalance auto-rebalance and proceed first with scaling operation. The order shoulder be -> scale-down operation, then scale-up and then imbalance rebalance. I will be making things more clear in the FSM diagram explanation

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,510 @@
+The operator will use the `unfixableAnomalies` list to check if there are anomalies with unfixable goals or not.
+If the list is not empty then the operator will create a `warning` condition based on the unfixable goal and add it to the `Kafka` CR status.
+
+In case we get an anomaly which has both fixable goals and unfixable goals, we will still `IGNORE` that anomaly. 
```
Anomaly detector will keep on detecting the unfixed anomaly until/unless the user fixes it. What I plan was that the operator will clear the `unfixableAnomaly` list no matter if it is resolved or not, which means that the the user will be keep getting the warning unless they decide to fix it because the detector will keeping detecting the unfixed anomaly. Once it is fixed the anomaly detector won't detect it again and since the operator already clears the ConfigMap every time, the warning will go away since the anomaly was fixed.

ShubhamRwt replied:
I will add a line about this in the proposal to make things clear

ShubhamRwt committed:
Added more changes

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Added suggestion by Tina and Kate

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

tinaselenge reviewed (approved):
Thanks @ShubhamRwt for addressing the comments. I am happy with the proposal, +1 non-binding

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,518 @@
+  fixableAnomalies: |
+    - <anomaly-id-1>:
+        - <fixable-goal-1>
+        - <fixable-goal-2>
```
Just to be sure I got it. Whenever Cruise Control an anomaly, it can bring more goal violations with it? Any example?

ppatierno replied:
Never mind, I found the answer by looking at the code. It's clearly possible by looking at here https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetails.java#L69

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,518 @@
+        - <unfixed-goal-1>
+        - <unfixed-goal-2>
+    - <anomaly-id-4>:
+        - <unfixed-goal-3> 
```
We are strictly separating `fixableAnomalies` from `unfixableAnomalies` so fixable vs unfixable goals but I have a question ... is it possible to have a fixable anomaly which brings some fixable goals but other unfixable goals? Because in such a case, the above schema would not work.

ShubhamRwt replied:
if an Anomaly has both fixable and unfixable goals, then I was planning to have the anomaly reflected in both the lists `fixableAnomalies` and `unfixableAnomalies`

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,518 @@
+It is possible that a new anomaly was ignored in the process but the anomaly detector will detect it again in case the rebalance didn't fix it so we don't have to worry about it.
+
+The above also holds true for the manual rebalances.
+If a user is running a manual rebalance, and some anomaly is detected while it is happening, we should still not add it to the `fixable` anomalies list.
```
Same question as above. Who is "we" in this sentence?

ppatierno commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,518 @@
+On each reconciliation, the following process will be used:
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
+2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check if there was any ConfigMap created with name `<cluster-name>-goal-violation-map` and whether the `fixableAnomalies` and `unfixableAnomalies` list is empty or not, then the `full` rebalance (imbalance mode) would be performed if the `unfixableAnomalies` list is empty and the `fixableAnomalies` list is not empty.
```
Same question as before, do we really care if the `unfixableAnomalies` list is empty? Shouldn't we rebalancing just because the `fixableAnomalies` list is not empty?

ppatierno reviewed (commented):
@ShubhamRwt we are almost there ... keep doing the nice work on it. I left few comments but by addressing them I will approve the proposal.

ShubhamRwt commented on 106-auto-rebalance-on-imbalanced-clusters.md:
````diff
@@ -0,0 +1,518 @@
+```
+
+Once the warning is issued, the operator will clear the `unfixableAnomalies` list.
+In case the user doesn't fix the anomaly, the `unfixableAnomalies` list will be populated again with the unfixable goal violation, and it will keep happening until the user fixes the anomaly manually.
````
Yes, sounds correct. I will make this change

ppatierno replied:
I see your point @katheris.
First, let me highlight that when an anomaly has unfixable goals, there is no rebalance going to happen so we can't clear the list "after a rebalance has happened".
At the same time we need anyway to clear the anomaly from the list because, unluckily, when the same anomaly is detected by Cruise Control later on, it will got assigned a different ID. So if we don't delete the previous one (which is actually the same anomaly but with a different ID), we'll get the ConfigMap growing indefinitely.
Right now I can't see a good solution unless not adding the warning in the Kafka CR but just a log into the Kafka operator that an anomaly was ignored because not fixable.

ShubhamRwt committed:
Added the changes suggested by Paolo

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt committed:
Fixed the missing change

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,520 @@
+## Motivation
+
+Currently, if the cluster is imbalanced, the user would need to manually rebalance the cluster by using the `KafkaRebalance` custom resource.
+With smaller clusters, it is feasible to fix things manually. However, for larger ones it can be very time-consuming, or just not feasible, to fix all the imbalances on your own.
```
What is `very time-consuming` about this for larger clusters? Isn't the process exactly the same for small and large cluster when you use `KafkaRebalance`? It just takes longer between the steps. But that does not mean the process itself is more time-consuming.

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,520 @@
+Every detected anomaly is placed in a priority queue and each of the anomalies have a priority value assigned to them.
+Since we are working with only goal related violations, every anomaly will have same priority.
+This means that detected anomalies will be fixed on a first come, first served basis.
+In situations where the anomaly is unfixable, due to issues like a lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy `ReplicaCapacityGoal`, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored and the user will be alerted via a condition in the Kafka CR status.
```
> and the user will be alerted via a condition in the Kafka CR status

This definitely deserves more details about how will you do it.

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,520 @@
+In case we get an anomaly which has both fixable goals and unfixable goals, we will still `IGNORE` that anomaly. 
+This is because Cruise Control won't be able to generate a proposal that can fix all the configured goals for rebalance.
+
+The created ConfigMap will persist when `imbalance` mode is enabled for auto-rebalance and will be deleted if the user decides to remove `imbalance` mode from `spec.kafka.cruiseControl.autoRebalance`.
```
Wait, above you said it would be deleted before the rebalance starts 🤔

scholzj commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,520 @@
+
+The advantages of using a single ConfigMap for all the detected anomalies versus one per anomaly are:
+1. It provides a single point of reference for all the anomalies
+2. When dealing with multiple Kafka clusters, the number of ConfigMap(s) created will be equal to the number of clusters deployed
```
I get why these two advantages matter. But I think the lifecycle of the config map as you describe it below is a mess and will not work well.

If nothing else, you should properly describe it and not have it spread through the text. Maybe it would work and is just badly described. But I wonder if it is really the case an whether a more event driven approach is needed.

scholzj replied:
> does the API have that kind of detail or does it have to go through the Notifier? Would we lose info like fixable vs unfixable?

I cannto answer that, but I guess in theory the notifier can have its own REST API for this?

ShubhamRwt replied:
This is how the Anomaly detector state looks like 
```
AnomalyDetectorState: {selfHealingEnabled:[BROKER_FAILURE, DISK_FAILURE, GOAL_VIOLATION, METRIC_ANOMALY, TOPIC_ANOMALY, MAINTENANCE_EVENT], selfHealingDisabled:[], selfHealingEnabledRatio:{BROKER_FAILURE=1.0, DISK_FAILURE=1.0, GOAL_VIOLATION=1.0, METRIC_ANOMALY=1.0, TOPIC_ANOMALY=1.0, MAINTENANCE_EVENT=1.0}, recentGoalViolations:[{anomalyId=c2071b83-b011-4924-8fa9-8d3cb0b2ebb9, detectionDate=2024-12-17T09:26:00Z, statusUpdateDate=2024-12-17T09:26:04Z, fixableViolatedGoals=[DiskUsageDistributionGoal], unfixableViolatedGoals=[], status=FIX_STARTED}], recentBrokerFailures:[], recentMetricAnomalies:[], recentDiskFailures:[], recentTopicAnomalies:[], recentMaintenanceEvents:[], metrics:{meanTimeBetweenAnomalies:{GOAL_VIOLATION:3.07 milliseconds, BROKER_FAILURE:0.00 milliseconds, METRIC_ANOMALY:0.00 milliseconds, DISK_FAILURE:0.00 milliseconds, TOPIC_ANOMALY:0.00 milliseconds}, meanTimeToStartFix:2.89 seconds, numSelfHealingStarted:1, numSelfHealingFailedToStart:0, ongoingAnomalyDuration=0.00 milliseconds}, ongoingSelfHealingAnomaly:None, balancednessScore:88.450}
```

The `recentGoalViolations` will have the anomalies even if they are resolved. I am not sure how we will differentiate if the violation is fixed or not. The only thing we could know is that whether a fix has started for the anomaly or not. I will try to see when this list gets cleared by running some tests but maybe the only way to clear the list would be when CC pod gets rolled again

tomncooper commented on 106-auto-rebalance-on-imbalanced-clusters.md:
```diff
@@ -0,0 +1,520 @@
+The `fixableAnomalies` list will state the `anomalyId` of the anomaly as well as the fixable violated goals.
+The `fixableAnomalies` list will be populated by the notifier every time it detects an anomaly.
+Whenever a rebalance is about to be triggered by the operator the `fixableAnomalies` list will be cleared.
+This will be the case for manual rebalances triggered by the user, as well as auto-rebalance rebalances that are happening in `add-broker`, `remove-broker`, or `imbalance` mode.
```
I agree there needs to be more discussion of how you will handle concurrent access and updates to the anomaly CM.

tomncooper reviewed (commented):
I have had another pass. I think the main issue is that, as Jakub highlighted, the lifecycle of the Anomaly CM is quite complicated.

I think the proposal would benefit from a description of the states the ACM will move through and what happens under different scenarios. When you have two process writing to and reading from the same source concurrently, you need to make sure there aren't any race conditions that could bite you. I suspect the relatively long anomaly detection period will help here but it is worth being methodical. 

I also think you may need a more sophisticated approach to the rebalance "completed with error" warnings in the Kafka CR status. You need to prevent thrashing or multiple warnings piling up for the same issue.

ShubhamRwt committed:
Updated the logic

Signed-off-by: ShubhamRwt <shubhamrwt02@gmail.com>

ShubhamRwt commented:
I will be closing this PR since we decided to change the logic more while addressing the comments on the PR and decided to not use any notifier at all. I will be creating a new PR for the same with the updated logic. Thanks for all the reviews
